File: server/src/scripts/applyMigrations.ts
import dotenv from "dotenv";
import fs from "fs/promises";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { isPostgresConfigured, pool } from "../lib/postgres.js";
import { logger } from "../lib/logger.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
dotenv.config({ path: resolve(__dirname, "../../../.env") });

async function ensureMigrationsTable() {
  await pool.query(`
    CREATE TABLE IF NOT EXISTS migrations (
      id SERIAL PRIMARY KEY,
      filename TEXT UNIQUE NOT NULL,
      checksum TEXT NOT NULL,
      applied_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
    );
  `);
}

async function computeChecksum(content: string): Promise<string> {
  const { createHash } = await import("crypto");
  return createHash("sha256").update(content).digest("hex").slice(0, 16);
}

async function applyMigrations() {
  if (!isPostgresConfigured()) {
    logger.error("Missing PostgreSQL configuration. Set DATABASE_URL in .env file.");
    process.exit(1);
  }

  try {
    await ensureMigrationsTable();
    logger.info("Connecting to PostgreSQL...");

    const filesToApply = [
      "01_users_colleges.sql",
      "02_questions.sql",
      "03_exams.sql",
      "04_jobs_pipeline.sql",
      "05_attempts_submissions.sql",
      "06_proctoring.sql",
      "07_interviews_feedback.sql",
      "08_platform_system.sql",
      "09_seed_data.sql",
      "10_indexes.sql",
      "11_audit_logs.sql",
      "12_soft_deletes.sql",
      "13_refresh_tokens.sql",
      "14_data_retention.sql",
      "15_slug_index.sql",
    ];

    for (const file of filesToApply) {
      const filePath = resolve(__dirname, `../../../database/${file}`);
      const sql = await fs.readFile(filePath, "utf-8");
      const checksum = await computeChecksum(sql);

      const existing = await pool.query(
        "SELECT checksum FROM migrations WHERE filename = $1",
        [file]
      );

      if (existing.rows.length > 0) {
        if (existing.rows[0].checksum === checksum) {
          logger.info(`Skipping already applied migration: ${file}`);
          continue;
        } else {
          logger.warn(`Migration ${file} has changed! Re-applying (checksum mismatch).`);
          // Remove old record to allow re-application
          await pool.query("DELETE FROM migrations WHERE filename = $1", [file]);
        }
      }

      logger.info(`Applying database file: ${file}...`);
      await pool.query(sql);
      await pool.query(
        "INSERT INTO migrations (filename, checksum) VALUES ($1, $2)",
        [file, checksum]
      );
      logger.info(`Successfully applied: ${file}`);
    }

    const countRes = await pool.query("SELECT COUNT(*)::int as count FROM questions");
    logger.info(`Verification: Total questions currently in database: ${countRes.rows[0].count}`);

  } catch (err: unknown) {
    logger.error({ err }, "Migration/Seeding failed");
    if (process.env.CI) {
      console.log(`::error::Migration/Seeding failed: ${err instanceof Error ? err.message + '\nStack: ' + err.stack : String(err)}`);
    }
    process.exit(1);
  } finally {
    await pool.end();
  }
}

applyMigrations();

File: server_py/scripts/apply_migrations.py
import os
import sys
import hashlib
import psycopg2
from dotenv import load_dotenv

# Setup import path for backend config and logging
base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(base_dir)

from app.config import DATABASE_URL
from app.logger import logger

def compute_checksum(content: str) -> str:
    h = hashlib.sha256()
    h.update(content.encode("utf-8"))
    return h.hexdigest()[:16]

def ensure_migrations_table(cur) -> None:
    cur.execute("""
        CREATE TABLE IF NOT EXISTS migrations (
            id SERIAL PRIMARY KEY,
            filename TEXT UNIQUE NOT NULL,
            checksum TEXT NOT NULL,
            applied_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
        );
    """)

def apply_migrations() -> None:
    if not DATABASE_URL:
        logger.error("Missing PostgreSQL configuration. Set DATABASE_URL in .env file.")
        sys.exit(1)

    logger.info("Connecting to PostgreSQL...")
    try:
        conn = psycopg2.connect(DATABASE_URL)
        cur = conn.cursor()
        
        ensure_migrations_table(cur)
        conn.commit()

        files_to_apply = [
            "01_users_colleges.sql",
            "02_questions.sql",
            "03_exams.sql",
            "04_jobs_pipeline.sql",
            "05_attempts_submissions.sql",
            "06_proctoring.sql",
            "07_interviews_feedback.sql",
            "08_platform_system.sql",
            "09_seed_data.sql",
            "10_indexes.sql",
            "11_audit_logs.sql",
            "12_soft_deletes.sql",
            "13_refresh_tokens.sql",
            "14_data_retention.sql",
            "15_slug_index.sql",
            "16_add_unique_exam_questions.sql",
        ]

        database_dir = os.path.join(os.path.dirname(base_dir), "database")

        for file in files_to_apply:
            file_path = os.path.join(database_dir, file)
            if not os.path.exists(file_path):
                logger.warning(f"Migration file not found: {file_path}")
                continue

            with open(file_path, "r", encoding="utf-8") as f:
                sql = f.read()

            checksum = compute_checksum(sql)

            cur.execute("SELECT checksum FROM migrations WHERE filename = %s", (file,))
            existing = cur.fetchone()

            if existing:
                if existing[0] == checksum:
                    logger.info(f"Skipping already applied migration: {file}")
                    continue
                else:
                    logger.warning(f"Migration {file} has changed! Re-applying (checksum mismatch).")
                    cur.execute("DELETE FROM migrations WHERE filename = %s", (file,))

            logger.info(f"Applying database file: {file}...")
            cur.execute(sql)
            cur.execute(
                "INSERT INTO migrations (filename, checksum) VALUES (%s, %s)",
                (file, checksum)
            )
            conn.commit()
            logger.info(f"Successfully applied: {file}")

        cur.execute("SELECT COUNT(*) as count FROM questions")
        total_questions = cur.fetchone()[0]
        logger.info(f"Verification: Total questions currently in database: {total_questions}")

        cur.close()
        conn.close()
        logger.info("Database migrations applied successfully!")

    except Exception as err:
        logger.error(f"Migration/Seeding failed: {err}")
        sys.exit(1)

if __name__ == "__main__":
    apply_migrations()
